src/charm.py

```diff
@@ -308,6 +311,11 @@ def _update_status(self, event):
         """Fired intermittently by the Juju agent."""
         self.unit.set_workload_version(self._prometheus_server.version())
 
+        # Call `_configure`` to resolve "BlockedStatus" when the ready-for-unit event is processed
+        # before ingress is in fact ready.
+        # https://github.com/canonical/traefik-k8s-operator/issues/78
+        self._configure(event)
```

Comment:
Copying comment by @rbarry82 from https://github.com/canonical/prometheus-k8s-operator/pull/340#pullrequestreview-1093589121:

> -   Update `IngressPerUnitRequirer` logic. This seems to be most of the changes in `src/charm.py`  
>     - Even there, I would suggest that "call `_configure|_common_exit_hook` hoping it will resolve `BlockedStatus` if an event was emitted before ingress is ready is a sledgehammer which is bad considering we already have an issue about `/-/reload` being called to often, what whatever logic it would take for "is `BlockedStatus` set during `_update_status`? Is it still valid? If it is, clear it, if not, move on) would be more explicit, cleaner, and more legible.
>     - Even seeing the comment and reading `_configure`, I have no idea why this would be true. Either `current_services == new_layer.services` and it's never restarted at all, so we never try to connect, or the change to `set_workload_version` by calling the binary rather than an endpoint means that the presumptive bypass by calling `self._prometheus_server.version()` will no longer call the endpoint, and this won't actually be set.
>     - In reading through this again, it also seems wrong that `# Reload may fail if ingress ...` wold lead to a `BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE)`. It's not corrupt, though. Ingress is not ready, or Prometheus has taken too long to come up. This could potentially lead administrators down a "blind alley"
>     - The "right" fix for this is likely:
>         - Add a `can_connect()` method to `prometheus_server.Prometheus()`. If `update_status` fires and it's `BlockedStatus` with a message (a _new_ message which indicates the actual state) about timeouts:
> 
> ```
> def _update_status(self, event):
>   ...
>   if isinstance(self.unit.status, BlockedStatus) and self.unit.status.message == PROMETHEUS_TIMEOUT:
>     if self._prometheus_server.can_connect():
>       self.unit.status = ActiveStatus()
> 
> ```
> - In addition, `prometheus_server.Prometheus.*` should provide distinctions between timeouts and `ConnectionError` to flag the "real" status. Custom exceptions instead of a boolean are a trivial, Pythonic way to do this

Reply:
> "call `_configure|_common_exit_hook` hoping it will resolve BlockedStatus if an event was emitted before ingress is ready is a sledgehammer which is bad considering we already have an issue about `/-/reload` being called to often

Agreed, but to limit the scope of this change, and given #342 will be addressed very soon, can live with this for now?

> what whatever logic it would take for "is BlockedStatus set during _update_status? Is it still valid? If it is, clear it, if not, move on) would be more explicit, cleaner, and more legible.

This would be in direct contrast to the common exit hook pattern. I'd rather not mix both approaches in the same charm. Also, once #342 is fixed this will be much less painful to look at.

> seems wrong that # Reload may fail if ingress ... wold lead to a `BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE)`

Agreed. I don't particularly like that status comparison pattern and it should be fixed as part of #342.

> prometheus_server.Prometheus.* should provide distinctions between timeouts and ConnectionError to flag the "real" status.

Sounds presumptuous at first glance. And what will the charm do with the various exceptions? Just choose one message over another? Maybe, if we have insight, expose a `retry` setting the charm could use?

Reply:
I mentioned this on #342.

Reply:
> > This would be in direct contrast to the common exit hook pattern. I'd rather not mix both approaches in the same charm. Also, once #342 is fixed this will be much less painful to look at.

I think that's a little bit the point -- the common exit hook cannot cleanly address this, but we actually do need to do so. There are multiple non-`_configure` places in the charm where we set `BlockedStatus` (in `_generate_command`, `_on_k8s_...`). "Un-setting" it isn't a massive divergence.

The odds of this being needed inside this patch are low, since ingress shouldn't fail/be unavailable long enough for us to want to avoid blocking a charm in `executing` for some timeout (2-5 seconds or whatever), though.
 
> Agreed. I don't particularly like that status comparison pattern and it should be fixed as part of #342.

Why would it be changed there? It also affects this. There seems to be three cases where the charm may not be accessible:

* Bad configuration (we can check with `promtool` if there's an exception raised trying to connect)
* Timeout because it's replaying/`mmap`ing HEAD chunks which haven't been written to the TSDB
* Not reachable due to ingress

`/-/reload` being called to often doesn't have anything to do with the last case, but this patch does.

> Sounds presumptuous at first glance. And what will the charm do with the various exceptions? Just choose one message over another? Maybe, if we have insight, expose a `retry` setting the charm could use?

Using exceptions for flow control is Pythonic, and a narrow custom exception which is caught and sets an appropriate message is much less presumptuous/magic than returning a boolean from `prometheus_server.Prometheus` which may set a meaningless message which will hopefully automagically clear itself the next time something fires and hits the common exit hook.

The exceptions could have a property with a message, or a lookup table in the charm, or whatever. The charm would set a status message to tell the administrator what is happening and why Prometheus may not be reachable which is more useful than a "configuration is invalid" message which may not be true or an `ActiveStatus` when Prometheus is actually not active/reachable.

A `retry` mechanism can only be so useful. For ingress, if we retry for 2-3 seconds, it may be enough. Traefik shouldn't take that long to restart. For `/-/reload`, though, it won't be, and how we address that can certainly be discussed elsewhere (we already "know" that we'll probably have to sidestep it with @manadart's work to trigger async hooks), but we're well enough aware of the "basic" failure cases, and we need a way to flag the administrator about what the actual status is. Since this patch is quite small/contained, and is the logical place to indicate what "goes wrong" when Ingress isn't available yet (timeout or default traefik landing page, I'm not sure which), the initial scaffolding of "raise an exception of the appropriate type" seems that it should go here and not be deferred.